File: src/2025-03-04-12-20-21--deepseek-vr-0304/cpp/exercises/practice/queen-attack/queen_attack.cpp
#include "queen_attack.h"
#include <cmath>

namespace queen_attack {

chess_board::chess_board(const std::pair<int, int>& white, const std::pair<int, int>& black)
    : white_pos(white), black_pos(black) {
    if (white_pos.first < 0 || white_pos.first >= 8 ||
        white_pos.second < 0 || white_pos.second >= 8 ||
        black_pos.first < 0 || black_pos.first >= 8 ||
        black_pos.second < 0 || black_pos.second >= 8) {
        throw std::domain_error("Position out of bounds");
    }
    if (white_pos == black_pos) {
        throw std::domain_error("Queens cannot be on the same square");
    }
}

std::pair<int, int> chess_board::white() const {
    return white_pos;
}

std::pair<int, int> chess_board::black() const {
    return black_pos;
}

bool chess_board::can_attack() const {
    if (white_pos.first == black_pos.first || white_pos.second == black_pos.second) {
        return true;
    }
    return std::abs(white_pos.first - black_pos.first) == 
           std::abs(white_pos.second - black_pos.second);
}

}  // namespace queen_attack

File: src/2025-03-04-03-28-40--deepseek-chat-0304/javascript/exercises/practice/queen-attack/queen-attack.js
export class QueenAttack {
  constructor({
    black: [blackRow, blackColumn] = [0, 3],
    white: [whiteRow, whiteColumn] = [7, 3],
  } = {}) {
    // Validate positions
    if (blackRow < 0 || blackRow > 7 || blackColumn < 0 || blackColumn > 7 ||
        whiteRow < 0 || whiteRow > 7 || whiteColumn < 0 || whiteColumn > 7) {
      throw new Error('Queen must be placed on the board');
    }
    
    // Check if queens are on same square
    if (blackRow === whiteRow && blackColumn === whiteColumn) {
      throw new Error('Queens cannot share the same space');
    }
    
    this.black = [blackRow, blackColumn];
    this.white = [whiteRow, whiteColumn];
  }

  toString() {
    const board = Array.from({length: 8}, () => Array(8).fill('_'));
    
    // Place queens
    board[this.black[0]][this.black[1]] = 'B';
    board[this.white[0]][this.white[1]] = 'W';
    
    // Convert to string representation
    return board.map(row => row.join(' ')).join('\n');
  }

  get canAttack() {
    const [blackRow, blackCol] = this.black;
    const [whiteRow, whiteCol] = this.white;
    
    // Check same row or column
    if (blackRow === whiteRow || blackCol === whiteCol) {
      return true;
    }
    
    // Check diagonals
    return Math.abs(blackRow - whiteRow) === Math.abs(blackCol - whiteCol);
  }
}
